spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerCompositeKeyProperty.java

```diff
@@ -0,0 +1,217 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.data.spanner.core.mapping;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.OptionalInt;
+
+import com.google.cloud.spanner.Key;
+
+import org.springframework.data.mapping.Association;
+import org.springframework.data.mapping.PersistentEntity;
+import org.springframework.data.util.ClassTypeInformation;
+import org.springframework.data.util.TypeInformation;
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+
+/**
+ * Represents an persistent property just to represent Spanner primary keys, and does not
+ * correspond to actual properties of POJOs, as it might even be a composite, multi-column key.
+ *
+ * @author Chengyuan Zhao
+ */
+public class SpannerCompositeKeyProperty implements SpannerPersistentProperty {
```

Comment:
What's the difference to a regular `SpannerPersistentProperty`? Using Spanner's `Key` as composite primary key seems a good fit.

We represented for quite a time multiple primary key columns in Spring Data Cassandra with a single property but at some point, this required more conditional code than following the simple approach of representing a composite key as regular persistent entity distinguished by the property of [`isCompositePrimaryKey()`](https://github.com/spring-projects/spring-data-cassandra/blob/1d61f3c56d12fb707663d49fb470036ac12eddfa/spring-data-cassandra/src/main/java/org/springframework/data/cassandra/core/mapping/BasicCassandraPersistentProperty.java#L297). We reverted the multiple-columns-in-property pattern with Spring Data Cassandra 2.0

Reply:
Regarding the difference between this and a regular SpannerPersistentPropertyImpl (SPPI), this one does not represent a real POJO property, and is never meant to be created or used anywhere other than as an ID property. It's also why SPPI's isIdProperty() is hard coded to give false, and this one true. Also as a result  ,because it's not a real property, has hardcoded simple implementations to return null or false to most of the PersistentProperty methods.

We also considered using the multi-columns-embedded-in-single ID class property approach, but felt it was less fitting for Spanner than a noSQL storage that has a more hierarchical data/table structure than Spanner. If were to use this "embedded" multiple-cols-in-single-property approach, we would ultimately need to flatten it into just a single table anyway, so I don't think we would be providing any real extra flexibility. 

Furthermore, I think there's real value in having the multiple primary key annotations in the POJO because it more closely mirrors what Spanner's table structure and syntax is.